plugins/@grouparoo/pardot/src/lib/export/exportProfiles.ts

```diff
@@ -0,0 +1,273 @@
+};
+
+// fetch using the keys to set destinationId and result on BatchExports
+// use the getByForeignKey to lookup results
+const findAndSetDestinationIds: BatchMethodFindAndSetDestinationIds = async ({
+  client,
+  foreignKeys,
+  getByForeignKey,
+}) => {
+  for (const email of foreignKeys) {
+    const prospect = await client.getProspectByEmail(email);
+    if (prospect) {
+      const user = getByForeignKey(email);
+      user.destinationId = prospect.id;
+      user.result = prospect;
+    }
+  }
+};
+
+// delete the given destinationIds
+const deleteByDestinationIds: BatchMethodDeleteByDestinationIds = async ({
+  client,
+  users,
+}) => {
+  for (const user of users) {
+    // No batch delete available, do one by one
+    await client.deleteProspectById(parseInt(user.destinationId));
+  }
+};
+
+// update these users by destinationId
+const updateByDestinationIds: BatchMethodUpdateByDestinationIds = async ({
+  client,
+  users,
+  config,
+}) => {
+  const { cacheData } = config.data;
+  await updateProspectsHelper(client, users, cacheData);
+};
+
+// usually this is creating them. ideally upsert. set the destinationId on each when done
+const createByForeignKeyAndSetDestinationIds: BatchMethodCreateByForeignKeyAndSetDestinationIds = async ({
+  client,
+  users,
+  config,
+}) => {
+  const { cacheData } = config.data;
+
+  await updateProspectsHelper(client, users, cacheData);
+
+  for (let user of users) {
+    // We're not using destinationIds for created prospects, but grouparoo requires it to be set
+    // and the batch upsert does not return IDs, so we'll just set it to the foreignKey
+    if (!user.destinationId) {
+      user.destinationId = user.foreignKeyValue;
+    }
+  }
+};
+
+async function updateProspectsHelper(
+  client: PardotClient,
+  users: BatchExport[],
+  cacheData: PardotCacheData
+) {
+  if (users.length === 0) return; // Nothing to do
+
+  // Sometimes we may get "414 - URI Too Long" from Pardot.
+  // In extremely long URIs, 414 won't be returned and "400 - Bad Request" will be returned instead.
+  // In these cases, we recursively split into two batches...
+  try {
+    await updateProspects(client, users, cacheData);
+  } catch (error) {
+    if (error.response?.status === 414 || error.response?.status === 400) {
+      if (users.length === 1) {
+        users[0].error = new Error(
+          `${error.response.status} - ${error.response.statusText}`
+        );
+        return;
+      }
+
+      const half = Math.ceil(users.length / 2);
+      await updateProspectsHelper(client, users.slice(0, half), cacheData);
+      await updateProspectsHelper(client, users.slice(half), cacheData);
+    } else {
+      throw error;
+    }
+  }
+}
+
+async function updateProspects(
+  client: PardotClient,
+  users: BatchExport[],
+  cacheData: PardotCacheData
+) {
+  const input: any[] = [];
+  for (const exportedProfile of users) {
+    const payload = await buildPayload(client, exportedProfile, cacheData);
```

Comment:
if there is an error (no email!)
this should catch and put the error jsut on that user.

Reply:
test for that too.

Reply:
`buildPayload` is not throwing any errors anymore based on previous comment, so this shouldn't be necessary.